File: backend/disease-service/src/main/java/disease_surveillance/disease_service/service/DiseaseService.java
package disease_surveillance.disease_service.service;

import disease_surveillance.disease_service.dto.DiseaseDto;

import java.util.List;
import java.util.Map;

public interface DiseaseService {

    // ── CRUD ──────────────────────────────────────────────────────────────────

    DiseaseDto.DiseaseResponse create(DiseaseDto.DiseaseRequest request);

    /** Returns a paginated list of all diseases. */
    DiseaseDto.PagedDiseaseResponse getAll(int page, int size, String sortBy, String sortDir);

    DiseaseDto.DiseaseResponse getById(Long id);

    DiseaseDto.DiseaseResponse update(Long id, DiseaseDto.DiseaseRequest request);

    void delete(Long id);

    // ── Lookups ───────────────────────────────────────────────────────────────

    DiseaseDto.DiseaseResponse getByName(String name);

    // ── Threshold queries ─────────────────────────────────────────────────────

    List<DiseaseDto.DiseaseResponse> getDiseasesByThreshold(Integer threshold);

    List<DiseaseDto.DiseaseResponse> getDiseasesAboveThreshold(Integer threshold);

    // ── Search ────────────────────────────────────────────────────────────────

    List<DiseaseDto.DiseaseResponse> searchDiseases(String keyword);

    // ── Aggregates / utility ──────────────────────────────────────────────────

    long getTotalDiseasesCount();

    boolean existsByName(String name);

    List<DiseaseDto.DiseaseResponse> getRecentDiseases();

    /** Returns a map of { diseaseName → thresholdLimit } for all diseases. */
    Map<String, Integer> getDiseaseThresholds();

    // ── Bulk operations ───────────────────────────────────────────────────────

    List<DiseaseDto.DiseaseResponse> bulkCreate(List<DiseaseDto.DiseaseRequest> requests);
}

File: frontend/src/api/diseaseService.js
import api from './axios'

// ── Disease Service ───────────────────────────────────────────
// Talks to: api-gateway → disease-service (Spring Boot, port 8083)
// All endpoints proxied through /disease-service/*

/**
 * GET /disease-service?page=0&size=10&sortBy=name&sortDir=asc
 * Returns: PagedDiseaseResponse { content, pageNumber, pageSize, totalElements, totalPages, last }
 */
export const getAllDiseases = (page = 0, size = 10, sortBy = 'name', sortDir = 'asc') =>
  api.get('/disease-service', { params: { page, size, sortBy, sortDir } })

/**
 * GET /disease-service/{id}
 * Returns: DiseaseResponse
 */
export const getDiseaseById = (id) =>
  api.get(`/disease-service/${id}`)

/**
 * GET /disease-service/name/{name}
 */
export const getDiseaseByName = (name) =>
  api.get(`/disease-service/name/${name}`)

/**
 * GET /disease-service/thresholds
 * Returns: { diseaseName: thresholdLimit } — used for dashboard Disease Distribution chart
 */
export const getDiseaseThresholds = () =>
  api.get('/disease-service/thresholds')

/**
 * GET /disease-service/count
 * Returns: number — total diseases registered
 */
export const getTotalDiseaseCount = () =>
  api.get('/disease-service/count')

/**
 * GET /disease-service/recent
 * Returns: DiseaseResponse[] — diseases added in last 30 days
 */
export const getRecentDiseases = () =>
  api.get('/disease-service/recent')

/**
 * GET /disease-service/search?keyword=fever
 */
export const searchDiseases = (keyword) =>
  api.get('/disease-service/search', { params: { keyword } })

/**
 * POST /disease-service
 * Body: DiseaseRequest { name, description, thresholdLimit }
 */
export const createDisease = (data) =>
  api.post('/disease-service', data)

/**
 * PUT /disease-service/{id}
 */
export const updateDisease = (id, data) =>
  api.put(`/disease-service/${id}`, data)

/**
 * DELETE /disease-service/{id}
 */
export const deleteDisease = (id) =>
  api.delete(`/disease-service/${id}`)

/**
 * POST /disease-service/bulk
 * Body: DiseaseRequest[]
 */
export const bulkCreateDiseases = (requests) =>
  api.post('/disease-service/bulk', requests)